File: server/c/glob/ActionClass.cpp
/* ***************************************************************************

   Copyright (c): 2013 Hericus Software, Inc.

   License: The MIT License (MIT)

   Authors: Steven M. Cherry

*************************************************************************** */

#include "ActionClass.h"
using namespace Helix::Glob;

static map_type* _ActionClassFactoryMap = NULL;

ActionClass* ActionClassFactory::createInstance(twine className, xmlNodePtr node) 
{
	map_type::iterator it = getMap()->find(className);
	if(it == getMap()->end()){
		DEBUG(FL, "ActionClassFactory::createInstance( %s , node) - no class found", className() );
		DEBUG(FL, "ActionClassFactory::map size: %d", getMap()->size() );
		throw AnException(0, FL, "ActionClassFactory::createInstance( %s , node) - no class found", className() );
	}
	return it->second(node);

}

map_type* ActionClassFactory::getMap()
{
	if(_ActionClassFactoryMap == NULL){
		_ActionClassFactoryMap = new map_type;
	}
	return _ActionClassFactoryMap;
}

ActionClass::~ActionClass()
{

}

bool ActionClass::isLongRunning()
{
	return false; // the default is to return false.
}

twine ActionClass::lrTaskName()
{
	return "Action Class - Generic Task Name";
}

File: server/net/glob/ActionClass.cs
/* ***************************************************************************

   Copyright (c): 2014 Hericus Software, Inc.

   License: The MIT License (MIT)

   Authors: Steven M. Cherry

*************************************************************************** */

using System;

namespace Helix.Glob
{
	/// <summary>
	/// The parent class for all of our logic actions that the server supports.
	/// </summary>
	public abstract class ActionClass
	{
		public string m_name;
		public string m_license_group;
		public bool m_requires_session;
		public bool m_delete_required;


		public ActionClass ()
		{
		}

		/// <summary>
		/// Must be implemented by children - this will do the work of actually executing
		/// the logic represented by this task.
		/// </summary>
		/// <param name="ioc">Ioc.</param>
		public abstract void ExecuteRequest (IOConn ioc);

		/// <summary>
		/// Returns whether or not this task is a long running task.  Long running
		/// tasks will return immediately to the caller and place themselves in a 
		/// background queue that the caller can then poll until the task is done.
		/// </summary>
		/// <returns><c>true</c>, if the task is long running, <c>false</c> otherwise.</returns>
		public bool isLongRunning ()
		{
			return false;  // The default is to return false
		}

		/// <summary>
		/// Returns the name of this task.
		/// </summary>
		/// <returns>The task name.</returns>
		public string lrTaskName()
		{
			return "Action Class - Generic Task Name";
		}

	}
}

